proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,754 @@
+{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"prod"},
+      {key:"version" value:"v1"}
+    ]
+    match_type: MATCH_TYPE_AND
+  }
+]}
+```
+
+This will work fine for older clients that are not yet sending the
+`version` key, because their dynamic parameters will not match the new
+variant's constraints.  However, newer clients that are sending dynamic
+parameters `{env=prod, version=v1}` will run into ambiguity: those
+parameters can match either of the above variants of the resource.
+
+This situation will be avoided via a best practice that all authoritative
+xDS servers should have **all variants of a given resource specify
+constraints for the same set of keys**.
+
+In order to make this work for the case where the server starts sending
+the constraint on the new key before all clients are sending it, we
+provide the `exists` matcher, which will allow the server to specify
+a default explicitly for clients that are not yet sending a new key.
+In this example, the server would actually have the following two
+variants:
+
+```textproto
+// Existing variant for older clients that are not yet sending the
+// version key.
+{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"prod"},
+      {key:"version" exists:{} invert:true}
+    ]
+    match_type: MATCH_TYPE_AND
+  }
+]}
+
+// New variant for clients sending the version key.
+{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"prod"},
+      {key:"version" value:"v1"}
+    ]
+    match_type: MATCH_TYPE_AND
+  }
+]}
+```
+
+This allows maintaining the requirement that all variants of a given
+resource have constraints on the same set of keys, while also allowing
+the server to explicitly provide a result for older clients that do not
+yet send the new key.
+
+##### Variants With Overlapping Constraint Values
+
+There is also a possible ambiguity that can occur if a server provides
+multiple variants of a resource whose constraints for a given key
+overlap in terms of the values they can match.  For example, let's say
+that a server has the following two variants of a resource:
+
+```textproto
+// Matches {env=prod} or {env=test}.
+{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"prod"},
+      {key:"env" value:"test"}
+    ]
+    match_type: MATCH_TYPE_OR
+  }
+]}
+
+// Matches {env=qa} or {env=test}.
+{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"qa"},
+      {key:"env" value:"test"}
+    ]
+    match_type: MATCH_TYPE_OR
+  }
+]}
+```
+
+Now consider what happens if a client subscribes with dynamic parameters
+`{env=test}`.  Those dynamic parameters can match either of the above
+variants of the resource.
+
+This situation will be avoided via a best practice that all authoritative
+xDS servers should have **all variants of a given resource specify
+non-overlapping constraints for the same set of keys**.  Control planes
+must not accept a set of resources that violates this requirement.
```

Comment:
Thinking out loud. Isn't this really up to the deployment? Meaning, this can be the recommendation, but wouldn't it also be fine to have some kind of internal ordering/priority that would kick in if there is an overlap?

Reply:
That might work for a simple deployment with no caching xDS proxies.  But as soon as you introduce caching proxies, you need to have the same ordering/priority logic in both the authoritative server and in all of the caching proxies, or else you'd get very non-deterministic results.  I don't think that would wind up being feasible.